packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -255,110 +298,140 @@ export class ChainService implements ChainServiceInterface {
-      ).subscribe({
-        next: async event => {
-          switch (event.type) {
-            case Deposited:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed Deposited event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.holdingUpdated(event);
-              break;
-            case AssetTransferred:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed AssetTransferred event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.assetTransferred(event);
-              break;
-            default:
-              throw new Error('Unexpected event from contract observable');
-          }
-        },
-      });
-      const subscriptions = this.channelToSubscription.get(channelId) ?? [];
-      this.channelToSubscription.set(channelId, [...subscriptions, subscription]);
+      // Fetch the current contract holding, and emit as an event
+      this.getInitialHoldings(contract, channelId).then(initialHoldings =>
+        subscriber.holdingUpdated(initialHoldings)
+      );
     });
   }
 
-  /** Implementation note:
-   *  The following is a simplified API that assumes a single registerChannel call per channel.
-   *  If we would like to allow for multiple registrations per channel, registerChannel should return a registration ID.
-   *  unregisterChannel will require the registration ID as a parameter.
-   */
   unregisterChannel(channelId: Bytes32): void {
-    const subscriptions = this.channelToSubscription.get(channelId);
-    if (subscriptions?.length !== 1) {
-      throw new Error(
-        'Unregister channel implementation only works when there is one subscriber per channel'
-      );
+    this.channelToSubscribers.set(channelId, []);
+  }
+
+  private async whenFinalized(channelId: Bytes32, finalizesAtS: number, tryNum: number) {
+    if (tryNum > 4) return;
+
+    const nowMs = Date.now();
+    const finalizesAtMs = finalizesAtS * 1_000;
+    if (nowMs < finalizesAtMs) {
+      setTimeout(() => this.whenFinalized(channelId, finalizesAtS, 0), nowMs - finalizesAtMs);
+      return;
     }
-    subscriptions.map(s => s.unsubscribe());
+    const [, storageFinalizesAtS] = await this.nitroAdjudicator.getChannelStorage(channelId);
+    const blockTimeS = (await this.provider.getBlock(await this.provider.getBlockNumber()))
+      .timestamp;
+    if (blockTimeS >= storageFinalizesAtS) {
+      // Should we wait for 6 blocks before emitting the finalized event?
+      // Will the wallet sign new states or deposit into the channel based on this event?
+      // The answer is likely no.
+      // So we probably want to emit this event as soon as possible.
```

Comment:
I think it's true that the _wallet_ might not, but it may emit an event that states that the channel's status is now `closed` officially, which the _app_ might take some action on.